README
AppleII-LibA2 version {{$version}}, released {{$date}}{{
  use Text::Tabs;

#=====================================================================
# Subroutines:
#---------------------------------------------------------------------
# Print a module name and current version number:

sub list_mod
{
    my $file = $_[0];
    my $path = "lib/$file";
    my ($f) = grep { $_->name eq $path } @{ $zilla->files };

    $f->content =~ /^\s*our\s+\$VERSION = '([0-9.]+)'/m
      or die "No version for $file";
    my $version = $1;

    $file =~ s!/!::!g;
    $file =~ s/\.pm$//;
    "\n    $file\tversion $version";
} # end list_mod

#---------------------------------------------------------------------
# Print a script name and current version number:

sub list_util
{
    my $file = $_[0];
    my $path = "bin/$file";
    my ($f) = grep { $_->name eq $path } @{ $zilla->files };

    $f->content =~ /^\s*our\s+\$VERSION = '([0-9.]+)'/m
      or die "No version for $file";
    my $version = $1;

    $file .= "\t" unless length($file) > 11;
    "\n    $file\tversion $version";
} # end list_util

  '';                           # insert nothing
}}

LibA2 provides both end-user utilities and library modules for
accessing files on Apple II ProDOS disk images for use with most Apple
II emulators.

This is a BETA release of LibA2.  There's some documentation, but it's
not complete.  If you want to see the rest of the documentation, then
send email!  Otherwise, I'll probably never get around to writing it.
Until then, use the source, Luke!  There are still probably some bugs,
and the interfaces might still change.  Use at your own risk.  Keep a
recent backup handy.  Wait sixty minutes before swimming.

All this having been said, I'm not aware of any serious bugs in LibA2
(besides the ones listed in the BUGS section below).  Good luck!
I would very much like to hear from everyone who tries LibA2.  The
more comments I get, the more likely I am to do more work on it.
Please send comments and questions to me.  Bug reports and patches
should go to the CPAN RT for LibA2: <bug-LibA2 AT rt.cpan.org>, or
through the web interface:

  http://rt.cpan.org/Public/Bug/Report.html?Queue=LibA2

You'll find my email address at the end of this file.


The included utilities are:
{{ list_util('prodos') }}
	`prodos' is the main end-user utility.  It provides a
	Unix-style shell for accessing ProDOS volumes.  This allows
	you to list the contents of disk images, create
	subdirectories, and copy files to & from disk images.  If you
	have installed the Term::ReadKey and Term::ReadLine modules,
	the shell will have better editing, command & filename
	completion, and a command history.

	The parameters for prodos are:
		prodos IMAGE_FILE

{{ list_util('pro_fmt') }}
	`pro_fmt' creates blank ProDOS disk images.  The images are
	NOT bootable, because they lack the necessary code in the boot
	blocks.  You can copy blocks from a bootable disk image to
	fix this.  The parameters for pro_fmt are:
		pro_fmt [options] IMAGE_FILE

{{ list_util('pro_opt') }}
	`pro_opt' removes unused space from ProDOS disk images.  This
	is most useful for reducing the size of hard disk images.  It
	doesn't use any form of compression; it simply moves
	everything to the beginning of the disk, squashing out empty
	space caused by deleting files.  Be careful with this, as it's
	likely to have some bugs left.  The parameters for pro_opt are:
		pro_opt SOURCE_IMAGE_FILE  DESTINATION_IMAGE_FILE

{{ list_util('awp2txt') }}
	`awp2txt' converts AppleWorks word processor files into text
	files.  This is a bit out of place in LibA2, because it has
	nothing to do with disk images, but I included it because
	users of LibA2 may find it useful.  The parameters for awp2txt are:
		awp2txt FILE ...

{{ list_util('var_display') }}
	`var_display' lists the contents of an Applesoft BASIC VAR file.
	Currently, it can only display string variables and string arrays.
	Like awp2txt, you must first extract the VAR file from the disk
	image.  The parameters for var_display are:
		var_display FILE


For people interested in writing their own utilities in Perl, LibA2
provides Perl 5 modules that supply classes for accessing ProDOS disk
images.  DOS 3.3 disks are not currently supported (except by
AppleII::Disk, which doesn't care what kind of data is on the disk).


The included modules are:
{{ list_mod('AppleII/Disk.pm') }}
	`AppleII::Disk' provides block-level access to disk images.
	It's useful because there are two main formats for Apple disk
	images:  ProDOS order and DOS 3.3 order.  These formats do not
	refer to the operating system used on the disk, but to the
	order in which the data appears.  AppleII::Disk takes care of
	the differences for you.

{{ list_mod('AppleII/ProDOS.pm') }}
	`AppleII::ProDOS' provides tools for accessing files on ProDOS
	disk images.  `prodos' is basically just a wrapper around
	AppleII::ProDOS.



INSTALLATION

To install this module, run the following commands:

{{ $t->build_instructions }}

If you do not want to install the utility scripts (AppleII:: modules
only), then use "perl Build.PL --no-scripts" instead of "perl Build.PL".



DEPENDENCIES

{{ $t->dependency_list }}



CHANGES
    Here's what's new in version {{$version}} of LibA2:
    (See the file "Changes" for the full revision history.)

{{$changes}}



BUGS
    Documentation
	There isn't enough.  Yet.  If you'd like to read it, send
	email.  If I think people want to read it, I'm far more likely
	to write it. :-)  Until then, use the source, Luke.

    prodos
	There's no way to copy a file from a disk image under another
	name.  This may cause problems under MS-DOS when extracting
	files that don't meet its 8.3 filename limitations.
	Solution: Upgrade to a real operating system.

    pro_fmt
	The disk images it creates aren't bootable unless you copy the
	boot blocks from a bootable disk.



AUTHOR

Christopher J. Madsen <perl AT cjmweb.net>



COPYRIGHT AND LICENSE

{{ local $_ = $zilla->license->notice; chomp; $_ }}


Changes
Revision history for AppleII-LibA2 (formerly LibA2)

0.201  NOT

	- Remove use of 2-argument open from scripts
	- Add 00-load.t

0.200  2015-08-01 beta

	- Distribution renamed from LibA2 to AppleII-LibA2
	- No functional changes

	[NEW MODULE]
	[AppleII::LibA2]
	- Placeholder module matching distribution name to comply with
	  the Lancaster Consensus

0.11   2013-12-26 beta

	- No functional changes
	- Release tests moved to xt/

0.10   2011-12-12 beta

	- Fixed Exporter prereq (need 5.57)
	- Remove IO::All from prereqs (picked up by mistake)

0.09   2011-11-11 beta

	- Require at least Perl 5.006

	[UPDATED UTILITIES]
	[prodos]
	- Use Term::ReadLine->Attribs to set up completion

	[var_display]
	- Version number reduced from 1.08 to 0.09

0.08   2007-03-24 beta

	- Updated my email address and some documentation.
	- Added POD tests.
	- Added another fix for the --no-scripts install option.
	- There are no functional changes.

0.07   2007-02-03 beta

	- I discovered that the --no-scripts install option was not working.
	  Only the Build.PL script was changed; no modules were updated.

0.06   2006-10-08 beta

	[AppleII::Disk]
	- The write_blocks method now treats block 0 specially

	[AppleII::ProDOS]
	- Can now write sparse files
	- pack_date now accepts Date::Calc format dates
	- Added unpack_date

0.05   2006-03-27 beta

	- Fixes for the build process (it didn't like spaces in pathnames)
	  If you got LibA2 0.04 to pass tests, you don't need to update.

	[AppleII::Disk]
	- Now compatible with versions of IO::File older than 1.11

0.04   2006-03-26 beta

	- Now uses Module::Build instead of ExtUtils::MakeMaker
	- You can use --no-scripts with Build.PL to avoid installing
	  utility scripts
	- Now has tests for AppleII::Disk & AppleII::ProDOS
	- Removed the .pl extension from utility scripts

	[UPDATED UTILITIES]
	[prodos]
	- Fixed to work with current Term::ReadLine::Perl
	- Added md as an alias for mkdir

	[pro_fmt]
	- Now has command line options (try --help for a list)

	[pro_opt]
	- Updated documentation

	[UPDATED MODULES]
	[AppleII::Disk]
	- INCOMPATIBLE CHANGE:
	  pad_block is now just a subroutine, not a method
	- Uses IO::File instead of FileHandle
	- The read_blocks method now treats block 0 specially

	[AppleII::ProDOS]
	- Can now read sparse files correctly
	- Fixed some places where filenames weren't being uppercased
	- Fixed a Y2K issue in converting timestamps to ProDOS format

0.003  1997-02-26 beta

	- First beta release of LibA2

	[NEW UTILITY]
	- var_display.pl	version 1.004

	[UPDATED UTILITIES]
	[awp2txt.pl version 0.005]
	- Includes documentation

	[prodos.pl version 0.010]
	- Includes documentation
	- Supports an external pager for long text displays

	[pro_opt.pl version 0.004]
	- Includes documentation
	- Now runs much faster

	[UPDATED MODULES]
	[AppleII::Disk version 0.009]
	- Includes documentation
	- Added blocks method to get/set disk size

	[AppleII::ProDOS version 0.026]
	- Includes documentation (still incomplete)
	- Adding a file to a full subdirectory automatically expands the
	  subdirectory.
	- Allocating free blocks is now much more efficient.	This makes
	  the volume optimizer (pro_opt.pl) much much faster.

0.002  1996-08-12 alpha

	- First public release of LibA2

	[NEW UTILITIES]
	- prodos.pl             version 0.008
	- pro_fmt.pl            version 0.003
	- pro_opt.pl            version 0.002
	- awp2txt.pl            version 0.004

	[NEW MODULES]
	- AppleII::Disk         version 0.008
	- AppleII::ProDOS       version 0.018

0.001  1996-08-12 alpha

	- A problem during uploading caused this version to be rechristened
	  version 0.002.


README.PL
#! /usr/bin/perl
#---------------------------------------------------------------------
# $Id$
# Copyright 1996 Christopher J. Madsen
#
# Create the README file for LibA2
#---------------------------------------------------------------------

use Text::Tabs;

# Read version number from Makefile.PL:

my $version;
open(PL, "<Build.PL") or die;
while (<PL>) {
    $version = $1, last if /^\s* dist_version\s* => \s* '(.+?)'/x;
}
close PL;
die unless defined $version;

# Generate README:

print "Generating README...\n";

open(STDOUT,'>README') or die;
binmode STDOUT;

print "LibA2\tRelease $version\n";
print <<'EOT';
Copyright 1996 Christopher J. Madsen

LibA2 provides both end-user utilities and library modules for
accessing files on Apple II ProDOS disk images for use with most Apple
II emulators.

This is a BETA release of LibA2.  There's some documentation, but it's
not complete.  If you want to see the rest of the documentation, then
send email!  Otherwise, I'll probably never get around to writing it.
Until then, use the source, Luke!  There are still probably some bugs,
and the interfaces might still change.  Use at your own risk.  Keep a
recent backup handy.  Wait sixty minutes before swimming.

All this having been said, I'm not aware of any serious bugs in LibA2
(besides the ones listed in the BUGS section below).  Good luck!
I would very much like to hear from everyone who tries LibA2.  The
more comments I get, the more likely I am to do more work on it.
Please send comments, questions, bug reports, and patches to me.
You'll find my email address at the end of this file.

EOT
#'

print 'The included utilities are:';

list_util('prodos');
print <<'EOT';
	`prodos' is the main end-user utility.  It provides a
	Unix-style shell for accessing ProDOS volumes.  This allows
	you to list the contents of disk images, create
	subdirectories, and copy files to & from disk images.  If you
	have installed the Term::ReadKey and Term::ReadLine modules,
	the shell will have better editing, command & filename
	completion, and a command history.

	The parameters for prodos are:
		prodos IMAGE_FILE
EOT
#`
list_util('pro_fmt');
print <<'';
	`pro_fmt' creates blank ProDOS disk images.  The images are
	NOT bootable, because they lack the necessary code in the boot
	blocks.  You can copy blocks from a bootable disk image to
	fix this.  The parameters for pro_fmt are:
		pro_fmt [options] IMAGE_FILE

#`
list_util('pro_opt');
print <<'';
	`pro_opt' removes unused space from ProDOS disk images.  This
	is most useful for reducing the size of hard disk images.  It
	doesn't use any form of compression; it simply moves
	everything to the beginning of the disk, squashing out empty
	space caused by deleting files.  Be careful with this, as it's
	likely to have some bugs left.  The parameters for pro_opt are:
		pro_opt SOURCE_IMAGE_FILE  DESTINATION_IMAGE_FILE

#`
list_util('awp2txt');
print <<'';
	`awp2txt' converts AppleWorks word processor files into text
	files.  This is a bit out of place in LibA2, because it has
	nothing to do with disk images, but I included it because
	users of LibA2 may find it useful.  The parameters for awp2txt are:
		awp2txt FILE ...

#`
list_util('var_display');
print <<'';
	`var_display' lists the contents of an Applesoft BASIC VAR file.
	Currently, it can only display string variables and string arrays.
	Like awp2txt, you must first extract the VAR file from the disk
	image.  The parameters for var_display are:
		var_display FILE

#`
print <<'EOT';

For people interested in writing their own utilities in Perl, LibA2
provides Perl 5 modules that supply classes for accessing ProDOS disk
images.  DOS 3.3 disks are not currently supported (except by
AppleII::Disk, which doesn't care what kind of data is on the disk).

EOT
#')

print "The included modules are:";

list_mod('AppleII/Disk.pm');
print <<'';
	`AppleII::Disk' provides block-level access to disk images.
	It's useful because there are two main formats for Apple disk
	images:  ProDOS order and DOS 3.3 order.  These formats do not
	refer to the operating system used on the disk, but to the
	order in which the data appears.  AppleII::Disk takes care of
	the differences for you.

#`
list_mod('AppleII/ProDOS.pm');
print <<'';
	`AppleII::ProDOS' provides tools for accessing files on ProDOS
	disk images.  `prodos' is basically just a wrapper around
	AppleII::ProDOS.

#'
print <<EOT;

CHANGES
    Here's what's new in version $version of LibA2:
    (See the file \`Changes\' for the full revision history.)
EOT

open(CHANGES,"<Changes") or die;
<CHANGES>;                      # Skip version on first line
while (<CHANGES>) {
    last if /^\*/;
    print unexpand(' 'x6 . expand($_));
}
close CHANGES;

print <<'EOT';
LICENSE
	LibA2 is free software; you can redistribute it and/or modify
	it under the same terms as Perl itself.

	LibA2 is distributed in the hope that it will be useful, but
	WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
	either the GNU General Public License or the Artistic License
	for more details.

BUGS
    Documentation
	There isn't enough.  Yet.  If you'd like to read it, send
	email.  If I think people want to read it, I'm far more likely
	to write it. :-)  Until then, use the source, Luke.

    prodos
	There's no way to copy a file from a disk image under another
	name.  This may cause problems under MS-DOS when extracting
	files that don't meet its 8.3 filename limitations.
	Solution: Upgrade to a real operating system.

    pro_fmt
	The disk images it creates aren't bootable unless you copy the
	boot blocks from a bootable disk.

AUTHOR
	Christopher J. Madsen <perl AT cjmweb.net>
EOT
#'

exit;

#=====================================================================
# Subroutines:
#---------------------------------------------------------------------
# Print a module name and current version number:

sub list_mod
{
    my $file = $_[0];
    my $version;
    open(PL,"<lib/$file") or die "Can't open lib/$file";
    while (<PL>) {
        $version = $1, last if /^\s*\$VERSION = '([0-9.]+)'/;
    }
    close PL;
    die "No version for $file" unless $version;
    $file =~ s!/!::!g;
    $file =~ s/\.pm$//;
    print "\n    $file\tversion $version\n";
} # end list_mod

#---------------------------------------------------------------------
# Print a script name and current version number:

sub list_util
{
    my $file = $_[0];
    my $version;
    open(PL,"<bin/$file") or die "Can't open bin/$file";
    while (<PL>) {
        $version = $1, last if /^our \$VERSION = '([0-9.]+)'/;
    }
    close PL;
    die "No version for $file" unless $version;
    $file .= "\t" unless length($file) > 11;
    print "\n    $file\tversion $version\n";
} # end list_util

# Local Variables:
#   tmtrack-file-task: "LibA2: README"
# End:
